libs/ledger-live-common/src/families/ethereum/hw-signMessage.ts

```diff
@@ -29,18 +34,60 @@ export const messageHash = (message: EIP712Message): Buffer => {
   );
 };
 
-const resolver: EthResolver = async (
+function tryConvertToJSON(message: string): string | EIP712Message {
+  try {
+    const parsedMessage = JSON.parse(message);
+    if (isEIP712Message(parsedMessage)) {
+      return parsedMessage as EIP712Message;
+    }
+  } catch {
+    // Not a JSON message
+  }
+  return message;
+}
+
+export const prepareMessageToSign = (
+  currency: CryptoCurrency,
+  path: string,
+  derivationMode: DerivationMode,
+  message: string
+): MessageData | TypedMessageData => {
+  const hexToStringMessage = Buffer.from(message, "hex").toString();
+  const parsedMessage = tryConvertToJSON(hexToStringMessage);
+
+  const messageData = {
+    currency,
+    path,
+    derivationMode,
+    rawMessage: "0x" + message,
+  };
+  if (typeof parsedMessage === "string") {
+    return {
+      ...messageData,
+      message: parsedMessage,
+    };
+  } else {
+    return {
+      ...messageData,
+      message: parsedMessage,
+      hashes: {
+        stringHash: "",
+        domainHash: bufferToHex(domainHash(parsedMessage)),
+        messageHash: bufferToHex(messageHash(parsedMessage)),
+      },
+    };
+  }
+};
+
+const signMessage: EthSignMessage = async (
   transport,
   { path, message, rawMessage }
 ) => {
   const eth = new Eth(transport);
-  const parsedMessage = (() => {
-    try {
-      return JSON.parse(message as string);
-    } catch (e) {
-      return message;
-    }
-  })();
+  let parsedMessage = message;
+  if (typeof message === "string") {
+    parsedMessage = tryConvertToJSON(message);
+  }
```

Comment:
`message` can be of the type `string` or `EIP712Message`. In the latter case, we can't and don't want to try to convert it into a JSON.
I added typing info in the method to be clearer.

Reply:
What I meant was to potentially change `tryConvertToJSON` into something like this:
```typescript
function tryConvertToEIP712(message: string | Record<string, any>): string | EIP712Message {
  const maybeParsedMessage = (() => {
    try {
      return JSON.parse(message as string) as Record<string, any>;
    } catch {
      return message;
    }
  })();
  
  return isEIP712Message(maybeParsedMessage)
    ? maybeParsedMessage 
    : message.toString();
}
```

So:
- if message is a random `string` it would throw on parse and return message as a `string`
- if message is a stringified random `object` it would fail on `isEIP712Message` and return message as a `string`
- if message is an `EIP712` stringified object it would return `maybeParsedMessage` as an `EIP712Message`
- if message is an `EIP712Message` it would throw on parse, succeed on `isEIP712Message` and return `maybeParsedMessage` as an `EIP712Message`
Only other adjustement to make would to add a type guard to `isEIP712Message` that the `message` param is an object first.

But again, I'm nit picking here, and it's just another abstraction, nothing major. 👍